Support formatting of subseconds

patrickroocks opened the pull request:
The formatting symbol `{%S}` should also print the subseconds, e.g., `fmt::format("{:%S}", std::chrono::microseconds(1234567))` should print "01.234567".

Didn't implement it for localized output as this localized output looks like black magic to me... may be we can omit it for the localized formatting for the first time?

patrickroocks commented:
Rebased on master.

vitaut commented on include/fmt/chrono.h:
```diff
@@ -966,13 +966,33 @@ inline void tzset_once() {
 #endif
 
-template <typename OutputIt, typename Char> class tm_writer {
+struct tm_ss {
+  std::tm tm;
+  int subsec_wdt;
+  uint64_t subsec;
+};
+
+struct empty_tm_ss {
+ private:
+  tm_ss tm_ss_;
+  empty_tm_ss() { tm_ss_.subsec_wdt = 0; }
+
+ public:
+  static empty_tm_ss& instance() {
+    static empty_tm_ss inst;
+    return inst;
+  }
+  const tm_ss& get_tm_ss() { return tm_ss_; }
+};
```
I don't think we need these. Just pass subseconds as an additional argument to `tm_writer`.

patrickroocks replied:
Removed and made subseconds an argumenbt.

vitaut commented on include/fmt/chrono.h:
```diff
@@ -966,13 +966,33 @@ inline void tzset_once() {
+  const tm_ss& get_tm_ss() { return tm_ss_; }
+};
+
+template <typename OutputIt, typename Char> class tm_ss_writer {
```
Let's keep the old name since formatting of subseconds is just a minor implementation detail.

patrickroocks replied:
Kept old name.

vitaut commented on include/fmt/chrono.h:
```diff
@@ -966,13 +966,33 @@ inline void tzset_once() {
   const std::locale& loc_;
   const bool is_classic_;
   OutputIt out_;
+  const tm_ss& tm_ss_;
```
Let's replace this with a variable holding subseconds to avoid duplication.

patrickroocks replied:
Done.

vitaut commented on include/fmt/chrono.h:
```diff
@@ -1135,10 +1155,18 @@ template <typename OutputIt, typename Char> class tm_writer {
 
  public:
-  tm_writer(const std::locale& loc, OutputIt out, const std::tm& tm)
+  tm_ss_writer(const std::locale& loc, OutputIt out, const tm_ss& tm_ss_val)
       : loc_(loc),
         is_classic_(loc_ == get_classic_locale()),
         out_(out),
+        tm_ss_(tm_ss_val),
+        tm_(tm_ss_val.tm) {}
+
+  tm_ss_writer(const std::locale& loc, OutputIt out, const std::tm& tm)
+      : loc_(loc),
+        is_classic_(loc_ == get_classic_locale()),
+        out_(out),
+        tm_ss_(empty_tm_ss::instance().get_tm_ss()),
         tm_(tm) {}
```
Similarly, there should be one ctor that takes an additional subseconds argument defaulted to 0.

patrickroocks replied:
I don't think it will work with an additional argument defaulted to 0. There is a difference between a timestamp based on `std::chrono::seconds(1)` (printed as "....01") and `std::chrono::milliseconds(1000)` (printed as "01.000" - the MS implementation of C++20 std::format does the same). Cf. the last unit test with `EXPECT_EQ(fmt::format("{:%S}", t10), "02.000");`

So I made two constructors and one sets `hasSubsecs` to true.

patrickroocks replied:
Used pointer now

vitaut commented on include/fmt/chrono.h:
```diff
@@ -2016,28 +2057,52 @@ struct formatter<std::chrono::time_point<std::chrono::system_clock, Duration>,
+              epoch - std::chrono::duration_cast<std::chrono::seconds>(epoch));
+
+      tm_ss_val.subsec = static_cast<uint64_t>(subsecs.count());
+      tm_ss_val.subsec_wdt =
```
What does wdt stand for?

patrickroocks replied:
width - removed anyway

vitaut commented on include/fmt/chrono.h:
```diff
@@ -2016,28 +2057,52 @@ struct formatter<std::chrono::time_point<std::chrono::system_clock, Duration>,
+    tm_ss_val.tm =
+        localtime(std::chrono::time_point_cast<std::chrono::seconds>(val));
+
+    constexpr bool reasonable_subsec_available =
+        Period::num == 1 &&
+        (Period::den == 1e1 || Period::den == 1e2 || Period::den == 1e3 ||
+         Period::den == 1e4 || Period::den == 1e5 || Period::den == 1e6 ||
+         Period::den == 1e7 || Period::den == 1e8 || Period::den == 1e9 ||
+         Period::den == 1e10 || Period::den == 1e11 || Period::den == 1e12 ||
+         Period::den == 1e13 || Period::den == 1e14 || Period::den == 1e15);
```
Can we support arbitrary fractions and not just 1/power-of-10?

patrickroocks replied:
Implicitly done.

vitaut commented on include/fmt/chrono.h:
```diff
@@ -2075,12 +2140,15 @@ template <typename Char> struct formatter<std::tm, Char> {
 
-  template <typename FormatContext>
-  auto format(const std::tm& tm, FormatContext& ctx) const
+  template <typename TimeType, typename FormatContext,
+            FMT_ENABLE_IF(std::is_same<TimeType, std::tm>::value ||
+                          std::is_same<TimeType, detail::tm_ss>::value)>
+  auto format(const TimeType& tm, FormatContext& ctx) const
```
Let's keep the first argument as `std::tm` and pass subseconds as an additional argument.

patrickroocks replied:
done

vitaut commented on test/chrono-test.cc:
```diff
@@ -652,3 +655,36 @@ TEST(chrono_test, utc_clock) {
+  auto t4_sub_sec_part = fmt::format("{0:09}", (t4 - t4_sec).count());
+  EXPECT_EQ(fmt::format("{}.{}", strftime_full(t4_sec), t4_sub_sec_part),
+            fmt::format("{:%Y-%m-%d %H:%M:%S}", t4));
+}
```
nit: missing newline

vitaut reviewed (changes requested):
Thanks for the PR.

patrickroocks commented:
I reworked everything again.

- The work regarding subseconds was already done in the `chrono_formatter` for formatting durations. I pulled the code out of that and put in helper functions `write_sub_sec` and `format_sec_fractional` used in the `tm_writer` and the `chrono_formatter`.
- This implicitly supports arbitrary ratios for the `Period`.
- I found a bug for printing durations with floating point types. The unit test 
`EXPECT_EQ(fmt::format("{:%S}", std::chrono::duration<double>{1.5}), "01.500000");`
didn't succeed, instead "01" was given as output. I changed the number of digits to at least 6, if the `Rep` type is fractional _and_ round(value) != value.

phprus commented on include/fmt/chrono.h:
```diff
@@ -966,13 +966,123 @@ inline void tzset_once() {
+
+// Format fractional seconds with an appropiate number of digits.
+template <typename Rep, long long Num, long long Den>
+basic_memory_buffer<char> format_sec_fractional(Rep val) {
```
Moving ``basic_memory_buffer`` may use copy (for buffers shorter than 500 chars).
See implementation:

https://github.com/fmtlib/fmt/blob/662adf4f33346ba9aba8b072194e319869ede54a/include/fmt/format.h#L864-L893

I think it's better to pass basic_memory_buffer by reference, like this:

https://github.com/fmtlib/fmt/blob/662adf4f33346ba9aba8b072194e319869ede54a/include/fmt/std.h#L50-L53

patrickroocks replied:
Ok thanks, didn't know that. Fixed.

phprus commented on include/fmt/chrono.h:
```diff
@@ -966,13 +966,123 @@ inline void tzset_once() {
   const bool is_classic_;
   OutputIt out_;
+  std::chrono::duration<Rep, Period> subsecs_;
+  bool hasSubsecs_;
```
Suggestion:
``has_subsecs_``

patrickroocks replied:
Fixed.

vitaut commented on include/fmt/chrono.h:
```diff
@@ -966,13 +966,122 @@ inline void tzset_once() {
+  static constexpr int value = (Num % Den == 0) ? N : 6;
+};
+
+// Format non-fractional subseconds with an appriate number of digits.
```
appriate -> appropriate

Also please keep the old name, `write_fractional_seconds`, and remove or update misleading comment.

patrickroocks replied:
done

vitaut commented on include/fmt/chrono.h:
```diff
@@ -966,13 +966,122 @@ inline void tzset_once() {
+};
+
+// Format non-fractional subseconds with an appriate number of digits.
+template <typename OutputIt, typename Duration, typename char_type>
```
char_type -> Char for consistency and make it the first argument so that the other ones are deduced.

vitaut commented on include/fmt/chrono.h:
```diff
@@ -966,13 +966,122 @@ inline void tzset_once() {
+  }
+}
+
+// Format fractional seconds with an appropiate number of digits.
```
Looks like the difference from the function above is that this one works with floating-point representation type rather than fractionality. Please update the comment accordingly.

patrickroocks replied:
done

vitaut replied:
This is better but I suggest changing to `write_floating_seconds` because we use `write` for consistency with `write_fractional_seconds`.

patrickroocks replied:
done

vitaut commented on include/fmt/chrono.h:
```diff
@@ -966,13 +966,122 @@ inline void tzset_once() {
+}
+
+// Format fractional seconds with an appropiate number of digits.
+template <typename Rep, long long Num, long long Den>
+void format_sec_fractional(basic_memory_buffer<char>& buf, Rep val) {
```
basic_memory_buffer<char> -> memory_buffer

Also pass duration directly instead of splitting it into components.

```c++
template <typename Rep, typename Period>
void format_sec_fractional(memory_buffer& buf, std::chrono::duration<Rep, Period> val) {
```

patrickroocks replied:
done

vitaut replied:
I'm pretty sure it should work as you can pass any rep type as part of duration type. This is already done in

```
        write_fractional_seconds<char_type>(
            out, std::chrono::duration<Rep, Period>(val));
```
so we should do the same here.

patrickroocks replied:
Ah, now I got it. I did exactly this, but used the `Rep` parameter and the tests failed for negative values... using `rep` as type it works fine.

vitaut commented on include/fmt/chrono.h:
```diff
@@ -966,13 +966,122 @@ inline void tzset_once() {
+  auto num_fractional_digits = count_fractional_digits<Num, Den>::value;
+  // For non-integer values, we ensure at least 6 digits to get microsecond
+  // precision.
+  if (num_fractional_digits < 6 && static_cast<Rep>(std::round(val)) != val) {
+    num_fractional_digits = 6;
+  }
```
Remove braces around the if body for consistency with the rest of the library.

patrickroocks replied:
Multiline-if without braces feels dangerous... I saw problematic bugs caused to this already sometimes.

But, of course, consistency first :)

vitaut commented on include/fmt/chrono.h:
```diff
@@ -966,13 +966,122 @@ inline void tzset_once() {
+            num_fractional_digits);
+}
+
+template <typename OutputIt, typename Char, typename Rep = int,
+          typename Period = std::ratio<1, 1>>
```
Let's replace Rep and Period with Duration for simplicity

patrickroocks replied:
ok, also replaced in
```
template <typename FormatContext, typename Duration>
  auto format(const std::tm& tm, FormatContext& ctx,
              const Duration& subsecs) const
```

vitaut commented on include/fmt/chrono.h:
```diff
@@ -1135,10 +1244,20 @@ template <typename OutputIt, typename Char> class tm_writer {
   }
 
  public:
+  tm_writer(const std::locale& loc, OutputIt out, const std::tm& tm,
+            std::chrono::duration<Rep, Period> subsecs)
```
Let's pass subsecs as a pointer with null indicating "no subseconds" and merge the two ctors into one to reduce copy pasta.

vitaut commented on include/fmt/chrono.h:
```diff
@@ -1337,9 +1456,28 @@ template <typename OutputIt, typename Char> class tm_writer {
+      if (has_subsecs_) {
+        if (std::is_floating_point<Rep>::value) {
+          auto buf = memory_buffer();
+          format_sec_fractional<Rep, Period::num, Period::den>(
+              buf, subsecs_.count());
```
subsecs should be passed as duration and template arguments deduced.

patrickroocks replied:
done

vitaut commented on include/fmt/chrono.h:
```diff
@@ -1337,9 +1456,28 @@ template <typename OutputIt, typename Char> class tm_writer {
+            out_ = std::copy(buf.begin() + 1, buf.end(), out_);
+          }
+        } else {
+          write_sub_sec<OutputIt, std::chrono::duration<Rep, Period>, Char>(
```
Only Char should be passed explicitly, the rest should be deduced once the argument order is changed per earlier comment.

patrickroocks replied:
done

vitaut commented on include/fmt/chrono.h:
```diff
@@ -2016,28 +2062,42 @@ struct formatter<std::chrono::time_point<std::chrono::system_clock, Duration>,
+          localtime(std::chrono::time_point_cast<std::chrono::seconds>(val)),
+          ctx, subsecs);
+
+    } else {
```
else is redundant, please remove

patrickroocks replied:
done

vitaut commented on include/fmt/chrono.h:
```diff
@@ -966,13 +966,124 @@ inline void tzset_once() {
+  return static_cast<Int>(value);
+}
+
+template <class Rep, class Period,
```
nit: class -> typename for consistency

patrickroocks replied:
done

vitaut commented on include/fmt/chrono.h:
```diff
@@ -966,13 +966,124 @@ inline void tzset_once() {
+
+// Format subseconds which are given as an integer type with an appropriate
+// number of digits.
+template <typename char_type, typename OutputIt, typename Duration>
```
char_type -> Char to follow naming conventions for template parameters

patrickroocks replied:
done

vitaut reviewed (changes requested):
Mostly looks good, just a few more minor comments.

vitaut commented:
Thank you!

vitaut commented:
Fuzzer discovered an issue with this change. The following code

```c++
#include <fmt/chrono.h>

int main() {
  fmt::format("{:%S}", std::chrono::duration<char, std::ratio<1, 100> >(0x80));
}
```
now results in an assertion failure:
```
    frame #7: 0x000000010000dddb a.out`fmt::v9::detail::assert_fail(char const*, int, char const*) + 59
    frame #8: 0x000000010000ac3e a.out`long long fmt::v9::detail::to_nonnegative_int<long long, long long, 0>(long long, long long) + 94
    frame #9: 0x000000010000a8dd a.out`void fmt::v9::detail::write_fractional_seconds<char, std::__1::back_insert_iterator<fmt::v9::basic_memory_buffer<char, 500ul, std::__1::allocator<char>>>, std::__1::chrono::duration<char, std::__1::ratio<1l, 100l>>>(std::__1::back_insert_iterator<fmt::v9::basic_memory_buffer<char, 500ul, std::__1::allocator<char>>>&, std::__1::chrono::duration<char, std::__1::ratio<1l, 100l>>) + 173
```

phprus commented:
@vitaut 
Fix: #3143

patrickroocks commented:
Thanks for the fix and sorry that I have overseen this!

vitaut commented:
Here's a link to the oss-fuzz report for future reference: https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=52380.
